Adding blog on Knative Vision (#5638)
knative/docs: blog/docs/articles/knative-eventing-vision.md

data.

The primary goal of secure event processing is to ensure the confidentiality, integrity, and
availability of the processed events while maintaining the privacy and security of the underlying
data.

### Transport encryption

Presently, event delivery within the cluster lacks encryption, imposing limitations on the types of
events that can be transmitted—usually those of low compliance value or with a relaxed compliance
posture. Alternatively, administrators must resort to a service mesh or encrypted CNI to encrypt the
traffic, introducing multiple challenges for Knative Eventing adopters.

The forthcoming solution lies in Knative Brokers and Channels, which will offer HTTPS endpoints to
receive events. However, since these endpoints typically lack public DNS names (e.g.,
svc.cluster.local or similar), they must be signed by a non-public CA, specific to the cluster or
organization.

To facilitate this, event producers will have the ability to know the CA root used to sign the
Broker or Channel certificate from the Knative resource's metadata.

For more detailed information, please refer to
the [transport-encryption feature](https://github.com/knative/docs/pull/5642)

> /* TODO use real feature docs link once PR above is in */

Reviewer: This is a TODO after https://github.com/knative/docs/pull/5642

# Secure Event processing and improved Event discoverability

**Authors: Pierangelo Di Pilato, Senior Software Engineer @ Red Hat, Matthias Weßendorf, Senior Principal Software Engineer @ Red Hat**

**Date: 2023-07-27**

## What is next in Knative Eventing?

> Knative Eventing has come a long way and has established itself as the goto platform on Kubernetes for event-driven applications. However we are not done with the project and we continue to evolve it. We are improving existing APIs and adding new features to Knative Eventing it as well. Below is a high level overview of what we are currently doing for the next couple of months.

Reviewer 2: ```suggestion
Knative Eventing has come a long way and has established itself as the goto platform on Kubernetes for event-driven applications. However we are not done with the project and we continue to evolve it. We are improving existing APIs and adding new features to Knative Eventing as well. Below is a high level overview of what we are currently doing for the next couple of months.
```

The primary goal of secure event processing is to ensure the confidentiality, integrity, and availability of the processed events while maintaining the privacy and security of the underlying data.

### Transport encryption

Currently, event delivery within the cluster is unencrypted. This limits the types of events which can be transmitted to those of low compliance value (or a relaxed compliance posture) or, alternatively, forces administrators to use a service mesh or encrypted CNI to encrypt the traffic, which poses many challenges to Knative Eventing adopters.

Knative Brokers and Channels will provide HTTPS endpoints to receive events. Given that these endpoints typically do not have public DNS names (e.g. svc.cluster.local or the like), these need to be signed by a non-public CA (cluster or organization specific CA).

Event producers will be able to determine the CA root used to sign the Broker or Channel certificate from the Knative resource’s metadata.

> For more information, please refer to the [feature track document](https://docs.google.com/document/d/1H-x_oji8LqkCyd7tlsSyclmUe7FAmEJPgRxOU_0pkn8/edit?resourcekey%253D0-lzDIPJsZOP3G17QE_g1lHw#heading%253Dh.8eg1on3fge8w).

Reviewer 2: ```suggestion
For more information, please refer to the [feature track document](https://docs.google.com/document/d/1H-x_oji8LqkCyd7tlsSyclmUe7FAmEJPgRxOU_0pkn8/edit).
```

### Sender identity

Currently, event delivery within the cluster is unauthenticated, and an event consumer cannot determine the identity of any sender.

By sending the identity of the event sender along with the events, event receivers can start implementing access control features.

For more information, please refer to the [feature track document](https://docs.google.com/document/d/1e7UgNTkL0Br5Da09Rg2ieVmhKJo4VXuBj-mHT9NCujY/edit).

### Authorization and admission policies

> Event-driven architecture is meant to lower organizations' silos and barriers, make systems more resilient and ultimately increase business agility, however, this is not possible without event admission policies and safeguards for security and data quality.

Reviewer 3: All the other sections have a section about what the current state is. Wanna write something similar?

Even if that we don't have anything, and brokers+channels+etc will simply accept any events.

Author: @pierDipi can you take a look here?

I think it is a good suggestion by Ali

Event admission policies will address the need to have authorization policies regarding who is allowed to send events to a given event hub, like a Knative Broker or Channel, as well as what they accept as valid content for events sent to them (schemas, etc).

Authorization and admission policies will be the feature that will bridge the secure event processing topic to the event discovery topic, below.

## Improved Event discovery

> Event discovery helps developers to understand which events they can consume and how they look like, without having to dig through various levels of service documentation.

Reviewer 2: ```suggestion
Event discovery helps developers to understand which events they can consume and how the events look like, without having to dig through various levels of service documentation.
```

Knative defines an EventType API which allows for discovery of the types of events that are available in the Knative Eventing system. This helps developers understand what kinds of events can be listened to and processed, which can be particularly helpful in systems with a large number of events being produced.

> However currently the EventType is unfortunately underused and is limited to Source-Duck compliant sources, in combination with the Knative Broker API. 

Reviewer 3: Any links to "Source-Duck compliant sources" ? I guess only 5-6 people will get what this means in the entire world :D

Author: But it is a fact, since that auto-create only works under some rules.

Any suggestion to say this differently?

Reviewer 2: Maybe something like 

"However currently the EventType is unfortunately underused and is limited to sources which are compliant with / implement the Knative [Source Ducktype](https://github.com/knative/pkg/blob/aef227e72ead466e1ff7a5dddca2c171b0d5f128/apis/duck/v1/source_types.go#L40), in combination with the Knative Broker API"

In order to improve the developer experience of Knative Eventing we are enhancing the event discovery in a couple of ways!

* Automatic event type creation
* Usage beyond the broker API
* Event Type definitions

### Automatic Event Type creation

Today the EventType API is limited to Source-Duck compliant sources, and only usable in combination with the Knative Broker API. 

> However when using the broker with 3rd party integration, such as regular deployments or even CLI tools like “kn event send” it does not work! We do not see any EvenType information for these common usage patterns.

Reviewer 2: ```suggestion
However when using the Broker with a 3rd party integration, such as regular deployments or even CLI tools like `kn event send` the EventType is not created automatically! We do not see any EvenType information for these common usage patterns.
```

> To address this problem we introduce an optional feature flag which can be ENABLED in order to have support for an automatic creation of EventTypes.

Reviewer 2: why ENABLED in upper case?

Author: no idea :sweat_smile:

Author: updated

Details can be found in the matching [feature track document](https://docs.google.com/document/d/1H8-mkMs5HWd3U7TT6KAWgU9ltDxqZv25Wls-6c4lneA/edit).

### Support for more than just Brokers

The EventType API will deprecate the “broker” field and replace it with a “reference”, which can be any Knative Resource, such a Channel, a Sink, or even a Database.

The good news is: This breaks the limitation to only use the event type metadata with the Broker API only. Even when there is no filter on Channel/Subscription APIs, it is still important to know what event types are emitted to components like channels or sinks!

### CRD for Event Type Definitions

> In addition to the above limitation, the fact that for EventType we need established connection between workloads (e.g. PingSource -> Broker) means It is not possible to tell what event type definitions are generally available in Knative Eventing, based on the installed sources/event emitters, independent of running workloads.

Reviewer 2: Not sure what you mean here / wording

> To address this problem, Knative Eventing will have a new Event Type Definition metadata CRD that can capture this information independently of running applications.

Reviewer 2: ```suggestion
To address this problem, Knative Eventing will have a new EventTypeDefinition metadata CRD that captures this information independently of running applications.
```

> Details can be found in the matching [feature track document]( https://docs.google.com/document/d/1vwEWtAm28g_QY9j0b63h8sRpGhvyB1K5ViNr8X3vIiM/edit).

Reviewer 2: ```suggestion
Details can be found in the matching [feature track document](https://docs.google.com/document/d/1vwEWtAm28g_QY9j0b63h8sRpGhvyB1K5ViNr8X3vIiM/edit).
```

Broker or Channel certificate from the Knative resource's metadata.

For more detailed information, please refer to

> the [transport-encryption feature](https://knative.dev/docs/eventing/experimental-features/transport-encryption/)

Reviewer: We need a relative file link here, not absolute URLs

Author: I've always used actual links for pointers to documentation :man_shrugging:

Reviewer 2: Link is not working in preview, but I think its, because you didn't rebase after #5642 merged.
/lgtm

/hold
to give others a chance to review too. Otherwise unhold...

Reviewer: Even the [eventtype-auto-creation feature documentation](https://deploy-preview-5638--knative.netlify.app/docs/eventing/experimental-features/eventtype-auto-creation.md) is not working

Reviewer: I'm not sure why tho, perhaps let's just go with direct links?

Reviewer 4: The blog and the docs are technically two different builds, so relative links won't work across them... in the nav, you need to use something like `/docs/eventing/experimental...` , but I don't recall offhand if that also works from the docs themselves

(there's some more info on the setup at https://github.com/knative/docs/tree/main/blog#docsblogconfignavyml)
